reflector.go

```diff
@@ -5,172 +5,203 @@ import (
-					"expected extra wrapping")
-			}
-		default:
-			return fmt.Errorf(
-				"unknown vault engine type: %q",
-				mapping.VaultEngineType,
-			)
+	// if we're not using the default label value, delete any secrets that are no longer in our
+	// mappings, but might still exist from previous runs in kubernetes
+	if r.labelValue != DefaultLabelValue {
+		if err := r.reconcile(ctx, r.secretsSet, touchedSecrets); err != nil {
+			return fmt.Errorf("error reconciling: %s", err)
 		}
+	}
 
-		// create the new Secret
-		newSecret := &v1.Secret{
-			ObjectMeta: metav1.ObjectMeta{
-				Name:      mapping.SecretName,
-				Namespace: r.k8sNamespace,
-				Labels: map[string]string{
-					LabelKey: r.labelValue,
-				},
-			},
-			Data: k8sSecretData,
-			Type: mapping.SecretType,
-		}
+	return nil
+}
+
+func (r *Reflector) getVaultSecret(mapping Mapping) (map[string][]byte, error) {
+	secretData, err := r.vaultClient.Read(mapping.VaultPath)
+	if err != nil {
+		return nil, fmt.Errorf("error reading vault key '%s': %s", mapping.VaultPath, err)
+	}
+
+	if secretData == nil {
+		return nil, fmt.Errorf("secret %s not found", mapping.VaultPath)
+	}
 
-		if _, ok := secretsSet[mapping.SecretName]; ok {
-			// secret already exists, so we should update it
-			_, err = secrets.Update(ctx, newSecret, metav1.UpdateOptions{})
+	// convert map[string]interface{} to map[string][]byte
+	var k8sSecretData map[string][]byte
+	switch mapping.VaultEngineType {
+	case vault.EngineTypeKeyValueV1:
+		k8sSecretData, err = r.castData(secretData.Data)
+		if err != nil {
+			return nil, fmt.Errorf("error casting data: %s", err)
+		}
+	case vault.EngineTypeKeyValueV2:
+		// there's an extra level of wrapping with the v2 kv secrets engine
+		if unwrapped, ok := secretData.Data["data"].(map[string]interface{}); ok {
+			k8sSecretData, err = r.castData(unwrapped)
 			if err != nil {
-				return fmt.Errorf("error updating secret: %s", err)
+				return nil, fmt.Errorf("error casting data: %s", err)
 			}
 		} else {
-			// secret doesn't exist, so create it
-			_, err = secrets.Create(ctx, newSecret, metav1.CreateOptions{})
-			if err != nil {
-				return fmt.Errorf("error creating secret: %s", err)
-			}
+			return nil, fmt.Errorf("key/value v2 interface did not have expected extra wrapping")
 		}
+	default:
+		return nil, fmt.Errorf("unknown vault engine type: %q", mapping.VaultEngineType)
+	}
 
-		log.Printf(
-			"reflected vault secret %s to kubernetes %s type (%s)",
-			mapping.VaultPath,
-			mapping.SecretName,
-			mapping.SecretType,
-		)
+	return k8sSecretData, nil
+}
 
-		// record the fact that we actually updated it
-		touchedSecrets[newSecret.Name] = struct{}{}
+func (r *Reflector) getGSMSecret(ctx context.Context, mapping Mapping) (map[string][]byte, error) {
+	resp, err := r.gsmClient.AccessSecretVersion(ctx, &secretmanagerpb.AccessSecretVersionRequest{
+		Name: mapping.GSMPath,
+	})
+	if err != nil {
+		return nil, err
```

Comment:
might be more descriptive to wrap this:
```suggestion
		return nil, fmt.Errorf("error accessing GSM secret %q: %w", mapping.GSMPath, err)
```